frontend/app/src/App.tsx

```diff
@@ -1038,17 +1027,14 @@ export class App extends PureComponent<Props, State> {
    * Handler called when the history state changes, e.g. `popstate` event.
    */
   onHistoryChange = (): void => {
-    const targetAppPage =
-      this.state.appPages.find(appPage =>
-        // The page name is embedded at the end of the URL path, and if not, we are in the main page.
-        // See https://github.com/streamlit/streamlit/blob/1.19.0/frontend/src/App.tsx#L740
-        document.location.pathname.endsWith("/" + appPage.pageName)
-      ) ?? this.state.appPages[0]
+    const { currentPageScriptHash } = this.state
+    const targetAppPage = this.appNavigation.findPageByUrlPath(
+      document.location.pathname
+    )
 
     // do not cause a rerun when an anchor is clicked and we aren't changing pages
     const hasAnchor = document.location.toString().includes("#")
-    const isSamePage =
-      targetAppPage?.pageScriptHash === this.state.currentPageScriptHash
+    const isSamePage = targetAppPage.pageScriptHash === currentPageScriptHash
```

Comment:
in the old version it looks like  `targetAppPage?` could have been `null` and the [next line ](https://github.com/streamlit/streamlit/pull/8744/files#diff-1b95d4e0dda1e5928f99c8303599bd811f140fdd40e028fcfca3077f802d6210R1039) still indicates that this might be possible. So should this be `targetAppPage?.pageScriptHash`?

Reply:
Yes, I may have removed it cause I forced the types to understand how MPA works. It might be safe to re-include this.